internal/provider/resource_eva_role.go

```diff
@@ -0,0 +1,270 @@
+		ScopedFunctionalities: data.getListOfFunctionalities(),
+	})
+
+	if attachPermissionsToRoleErr != nil {
+		resp.Diagnostics.AddError("Creating role permissions failed.", fmt.Sprintf("Unable to create role permisions, got error: %s", attachPermissionsToRoleErr))
+		return
+	}
+
+	tflog.Trace(ctx, "Created a new role.")
+
+	diags = resp.State.Set(ctx, &data)
+	resp.Diagnostics.Append(diags...)
+}
+
+func (r roleProdiver) Read(ctx context.Context, req tfsdk.ReadResourceRequest, resp *tfsdk.ReadResourceResponse) {
+	var data roleProviderTypeData
+
+	diags := req.State.Get(ctx, &data)
+	resp.Diagnostics.Append(diags...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	roleData, err := r.provider.evaClient.GetRole(ctx, eva.GetRoleRequest{
+		ID: data.ID.Value,
+	})
+
+	if err != nil {
+		resp.Diagnostics.AddError("Getting role unit failed.", fmt.Sprintf("Unable to get role, got error: %s", err))
+		return
+	}
+
+	data.Name = types.String{Value: roleData.Result.Name}
+	data.UserType = types.Int64{Value: roleData.Result.UserType}
+	data.Code = types.String{Value: roleData.Result.Code}
+	data.setListOfFunctionalities(roleData.Result.ScopedFunctionalities)
+
+	diags = resp.State.Set(ctx, &data)
+	resp.Diagnostics.Append(diags...)
+}
+
+func (r roleProdiver) Update(ctx context.Context, req tfsdk.UpdateResourceRequest, resp *tfsdk.UpdateResourceResponse) {
+	var data roleProviderTypeData
+
+	diags := req.Plan.Get(ctx, &data)
+	resp.Diagnostics.Append(diags...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	_, err := r.provider.evaClient.UpdateRole(ctx, eva.UpdateRoleRequest{
+		ID:       data.ID.Value,
+		Name:     data.Name.Value,
+		UserType: data.UserType.Value,
+		Code:     data.Code.Value,
+	})
+
+	if err != nil {
+		resp.Diagnostics.AddError("Updating role unit failed.", fmt.Sprintf("Unable to update role, got error: %s", err))
+		return
+	}
+
+	roleData, getRoleErr := r.provider.evaClient.GetRole(ctx, eva.GetRoleRequest{
+		ID: data.ID.Value,
+	})
+
+	if getRoleErr != nil {
+		resp.Diagnostics.AddError("Updating role permissions failed.", fmt.Sprintf("Unable to get role, got error: %s", getRoleErr))
+		return
+	}
+
+	_, detachErr := r.provider.evaClient.DetachFunctionalitiesFromRole(ctx, eva.DetachFunctionalitiesFromRoleRequest{
+		RoleID:                data.ID.Value,
+		ScopedFunctionalities: roleData.Result.ScopedFunctionalities,
+	})
+
+	if detachErr != nil {
+		resp.Diagnostics.AddError("Updating role permissions failed.", fmt.Sprintf("Unable to detach current role permissions, got error: %s", detachErr))
+		return
+	}
+
+	_, attachErr := r.provider.evaClient.AttachFunctionalitiesToRole(ctx, eva.AttachFunctionalitiesToRoleRequest{
+		RoleID:                data.ID.Value,
+		ScopedFunctionalities: data.getListOfFunctionalities(),
+	})
+
+	if attachErr != nil {
+		resp.Diagnostics.AddError("Updating role permissions failed.", fmt.Sprintf("Unable to attach new role permissions, got error: %s", detachErr))
+		return
+	}
```

Comment:
hmmmm... Now I see what you meant earlier. Remove everything and re-add everything 🤔